homeassistant/components/media_player/plex.py

```diff
@@ -272,93 +527,383 @@ def _active_media_plexapi_type(self):
+                    return (SUPPORT_PAUSE | SUPPORT_PREVIOUS_TRACK |
+                            SUPPORT_NEXT_TRACK | SUPPORT_STOP |
+                            SUPPORT_VOLUME_SET | SUPPORT_PLAY | SUPPORT_SEEK |
+                            SUPPORT_TURN_OFF)
+                # Only supports play,pause,stop (and off which really is stop)
+                elif self.make.lower().startswith("tivo"):
+                    _LOGGER.debug(
+                        'Tivo client detected, only enabling pause, play, '
+                        'stop, and off controls: %s', self.entity_id)
+                    return (SUPPORT_PAUSE | SUPPORT_PLAY | SUPPORT_STOP |
+                            SUPPORT_TURN_OFF)
+                # Default to show all controls
+                elif self.device:
+                    return (SUPPORT_PAUSE | SUPPORT_PREVIOUS_TRACK |
+                            SUPPORT_NEXT_TRACK | SUPPORT_STOP |
+                            SUPPORT_VOLUME_SET | SUPPORT_PLAY | SUPPORT_SEEK |
+                            SUPPORT_TURN_OFF | SUPPORT_VOLUME_MUTE)
+                else:
+                    return None
+
+    def local_client_control_fix(self):
+        """Detect if local client and adjust url to allow control."""
+        if self.device:
+            # if this device's machineIdentifier matches an active client
+            # with a loopback address, the device must be local or casting
+            for client in self.device.server.clients():
+                if "127.0.0.1" in client.baseurl:
+                    if (client.machineIdentifier ==
+                            self.device.machineIdentifier):
+                        # point controls to server since that's where the
+                        # playback is occuring
+                        _LOGGER.debug(
+                            'Local client detected, redirecting controls to '
+                            'Plex server: %s', self.entity_id)
+                        server_url = self.device.server.baseurl
+                        client_url = self.device.baseurl
+                        self.device.baseurl = "{}://{}:{}".format(
+                            urlparse(client_url).scheme,
+                            urlparse(server_url).hostname,
+                            str(urlparse(client_url).port))
 
     def set_volume_level(self, volume):
         """Set volume level, range 0..1."""
-        self.device.setVolume(int(volume * 100),
-                              self._active_media_plexapi_type)
+        if self.device:
+            self.local_client_control_fix()
+            self.device.setVolume(
+                int(volume * 100), self._active_media_plexapi_type)
+            self._volume_level = volume  # store since we can't retrieve
+
+    @property
+    def volume_level(self):
+        """Return the volume level of the client (0..1)."""
+        if self.state not in (STATE_IDLE, STATE_OFF):  # true or frozen state
```

Comment:
We should never have two embedded `if` statements. Instead they should be merged into 1 if statement.